pkg/client/unversioned/discovery_client.go

```diff
@@ -0,0 +1,133 @@
+type DiscoveryClient struct {
+	http.Client
+	config Config
+}
+
+// ServerAPIVersions returns the GroupVersions supported by the API server.
+// It creates a RESTClient based on the passed in config, but it doesn't rely
+// on the Version, Codec, and Prefix of the config, because it uses AbsPath and
+// takes the raw response.
+func (d *DiscoveryClient) ServerAPIVersions() (groupVersions []string, err error) {
+	client := d.Client
+
+	if d.config.Version != "" || d.config.Prefix != "" {
+		return nil, fmt.Errorf("The config of a DiscoveryClient should have empty version and prefix")
+	}
+	baseURL, err := defaultServerUrlFor(d.config)
+	if err != nil {
+		return nil, err
+	}
+	// Get the groupVersions exposed at /api
+	baseURL.Path = "/api"
+	resp, err := client.Get(baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	var v unversioned.APIVersions
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&v)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+
+	groupVersions = append(groupVersions, v.Versions...)
+	// Get the groupVersions exposed at /apis
+	baseURL.Path = "/apis"
+	resp2, err := client.Get(baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	var apiGroupList unversioned.APIGroupList
+	defer resp2.Body.Close()
+	err = json.NewDecoder(resp2.Body).Decode(&apiGroupList)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+	groupVersions = append(groupVersions, extractGroupVersions(&apiGroupList)...)
+
+	return groupVersions, nil
+}
+
+func (d *DiscoveryClient) SupportedResourcesForGroupVersion(groupVersion string) (*api.APIResourceList, error) {
+	var prefix string
+	baseURL, err := defaultServerUrlFor(d.config)
+	if groupVersion == "v1" {
+		baseURL.Path = "/api/" + groupVersion
+	} else {
+		baseURL.Path = "/apis/" + groupVersion
+	}
+	body, err := d.Client.Get(baseURL.String)
+	if err != nil {
+		return nil, err
+	}
+	resources := api.APIResourceList{}
+	if err := json.Unmarshal(body, &resources); err != nil {
+		return nil, err
+	}
+	return &resources, nil
+}
+
+// SupportedResources gets all supported resources for all group versions.  The key in the map is an API groupVersion.
+func (d *DiscoveryClient) SupportedResources() (map[string]*api.APIResourceList, error) {
+	apis, err := d.ServerAPIVersions()
+	if err != nil {
+		return nil, err
+	}
+	result := map[string]*api.APIResourceList{}
+	for _, groupVersion := range apis {
+		resources, err := d.SupportedResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return nil, err
+		}
+		result[groupVersion] = resources
+	}
+	return result, nil
+}
+
+func setDiscoveryDefaults(config *Config) error {
+	config.Prefix = ""
+	config.Version = ""
+	return nil
+}
+
+// NewDiscovery creates a new DiscoveryClient for the given config. This client
+// can be used to discover supported resources in the API server.
+func NewDiscovery(c *Config) (*ExtensionsClient, error) {
```

Comment:
Change return value :)

Also perhaps spell out the name, NewDiscoveryClient(...)